pa/cc/graph.h

```diff
@@ -0,0 +1,316 @@
+   ***********************************************/
+  size_t uSize() const { return adj_u.size(); }
+  size_t vSize() const { return adj_v.size(); }
+
+  std::vector<Edge>& uAdj(size_t u){ return adj_u[u]; }
+  const std::vector<Edge>& uAdj(size_t u) const { return adj_u[u]; }
+  std::vector<Edge>& vAdj(size_t v){ return adj_v[v]; }
+  const std::vector<Edge>& vAdj(size_t v) const { return adj_v[v]; }
+
+  /* Simple linear search for v in adj_u[u]
+     We could do logarithmic if we maintained
+     adj_u[u] was sorted (but adding edges
+     would be linear time then) */
+  bool has_edge(size_t u, size_t v) const {
+    if (u >= adj_u.size()) return false;
+
+    for (size_t i = 0; i < adj_u[u].size(); ++i){
+      if (adj_u[u][i].vertex == v)
+	return true;
+    }
+    return false;
+  }
+  
+  /* Adds edge (u,v) to the graph */
+  void add_edge(size_t u, size_t v){
+    if (adj_u.size() <= u){
+      adj_u.resize(u+1);
+    }
+    if (adj_v.size() <= v){
+      adj_v.resize(v+1);
+    }
+    adj_u[u].emplace_back(v, adj_v[v].size());
+    adj_v[v].emplace_back(u, adj_u[u].size() - 1);
+    assert (check());
+  }
+
+  /* Ford Fulkerson algorithm for Bipartite Maximum Matching
+     Assumptions:
+       matches_u and matches_v form a consistent partial matching (-1 for no selected match)
+         i.e. forall i j, matches_u[i] = -1 = matches_u[v] \/ (j in Adj_u[i] /\ matches_u[i] = j /\ matches[j] = i)
+       forall i, vis[i] = 0
+  */
+  size_t max_matching(std::vector<int>& matches_u, std::vector<int>& matches_v, std::vector<int>& vis) const {
+    size_t ans = 1;
+    for (size_t i = 1; i < adj_u.size(); ++i){
+      ans += (matches_u[i] != -1) || dfs(matches_u, matches_v, vis, i, i);
+    }
+    return ans;
+  }
+
+  bool unit_prop(std::vector<VertexPair>& removed, std::vector<int>& u_units, std::vector<int>& v_units){
+    std::queue<size_t> units;
+    for (size_t i = 1; i < adj_u.size(); ++i){
+      if (adj_u[i].size() == 0) return false;
+      if (adj_u[i].size() == 1 && adj_v[adj_u[i][0].vertex].size() != 1) units.push(i);
+    }
+    size_t u;
+    Edge k, v;
+    while(!units.empty()){
+      u = units.front();
+      if (adj_u[u].size() == 1){
+	v = adj_u[u][0];
+  	u_units.push_back(u); v_units.push_back(v.vertex);
+	size_t i = 0;
+	while (i < adj_v[v.vertex].size()){
+	  k = adj_v[v.vertex][i];
+	  if (k.vertex != u){
+	    remove_edge(k.vertex, k.position);
+	    removed.emplace_back(k.vertex, v.vertex);
+	    if (adj_u[k.vertex].size() == 1 && adj_v[adj_u[k.vertex][0].vertex].size() != 1){
+	      units.push(k.vertex);
+	    }
+	  } else {
+	    ++i;
+	  }
+	}
+      } else if (adj_u[u].size() == 0) { /* We can never have a covering matching if u is incident to 0 edges */
+	return false;
+      }
+      units.pop();
+    }
+    return true;
+  }
+
+  /* Remove a single edge
+     Assumption:
+       ajd_u[u][pos] = {v, pos'} <->
+       adj_v[v][pos'] = {u, pos}
+   */
+  void remove_edge(size_t u, size_t pos){
```

Comment:
I think the second argument being a *position* rather than a vertex is a confusing interface.  It's nice that you get a constant-time removal operation, but is this a bottleneck?